### PR TITLE
feat: generate a form where you can add validations to the field

### PR DESCRIPTION
**Pull request by JonasVanHoof**
Issue: [SBF-3](https://binnenland.atlassian.net/browse/SFB-3)

### CODE REVIEW

**Comment by JonasVanHoof** [comment]
| Done    | Todo                                                                                        | Comment                                                                                                                 |
| ------- | ------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------- |
| &check; | Validation is added correctly                                                               |                                                                                                                         |
| &check; | Validation is removed correctly                                                             |                                                                                                                         |
| &cross; | validation can be edited correctly                                                          | When assigning a required constrained clicking in the preview field and going out it adds the rcross that this is wrong |
| &cross; | You cannot add duplicate validations                                                        | You can add dupklicate validations. They are shown twice in the preview                                                 |
| &cross; | You can only add the validation when it is possible to add to the field (canHaveValidation) | Not implememted                                                                                                         |
| &cross; | You can only add the validation when it is possible to add to the field (canHaveValidation) | Not implememted                                                                                                         |

**Comment by JonasVanHoof** [comment]
@Windvis this PR should go about the `requiredConstrained` but the `maxLength` is also in here as we wahere not able to add the validations and show the fields when no conditional field was visible, so you will also find code for the maxLength validation.

The issue when selecting the maxLength is now that the red border is shown. This is because the grouping is still `Bag` instead of `MatchEvery`. After this is merged I can look at updating the grouping types for each validation in the form to the ones defined in the `validations/meta.ttl`

**Comment by JonasVanHoof** [comment]
Also the idea was to be able to pass enum values to t `a validate` method, the will than check all passed validations. But as in our usecase it should always validate every validation in the form so this can wait.. Will be improving the `areValidationsInGraphValidated` while adding the validator checks for upcoming validations.

File: .gitignore
Comment (Windvis) [inline .gitignore:5]: nitpick: Git has an option to use a ["global gitignore"](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file). That is where I add things specific to my development environment. I wouldn't add it here.

File: app/components/add-validations-to-form.js
Comment (Windvis) [inline app/components/add-validations-to-form.js:13]: Nitpick: it's usually better to use derived state if possible. Either by moving the logic to an ember-concurrency task (which has a running state) or by using a getter that returns true if builderStore and builderForm aren't falsy.

File: app/components/add-validations-to-form.js
Comment (Windvis) [inline app/components/add-validations-to-form.js:29]: Is this needed? I would use the args if there is no specific reason.

File: app/components/add-validations-to-form.js
Comment (Windvis) [inline app/components/add-validations-to-form.js:31]: I would move this to an async method or an ember-concurrency task and simply start it in the constructor.

File: app/components/add-validations-to-form.js
Comment (Windvis) [inline app/components/add-validations-to-form.js:18]: nitpick: What's up with the snake_case naming? 😄

File: app/components/add-validations-to-form.js
Comment (Windvis) [inline app/components/add-validations-to-form.js:66]: Nitpick: the name is a bit confusing since nothing is created, really. It's retrieved from the graph.

File: app/components/add-validations-to-form.hbs
Comment (Windvis) [inline app/components/add-validations-to-form.hbs:7]: Does this need to be set to `true`? This forces the error state for fields even if the user didn't touch them yet.

File: app/styles/_shame.scss
Comment (Windvis) [inline app/styles/_shame.scss:55]: I guess this belongs in the `c-playground.scss` file?

File: app/styles/project/_c-playground.scss
Comment (Windvis) [inline app/styles/project/_c-playground.scss:44]: Nitpick, I would rename the class to `.playground-tab`. It's not a child element of the `playground` block, so it feels weird to use the element notation here.

File: app/templates/formbuilder/edit.hbs
Comment (Windvis) [inline app/templates/formbuilder/edit.hbs:15]: The toggle is a bit weird since you can only click the "validation" label to toggle it. Not sure what the fix for that would be though so something for the UX team in one of the future sprints.

**Comment by JonasVanHoof** [comment]
If this can be merged we can check if we want to add the validation concept scheme selector as followed https://github.com/lblod/frontend-form-builder/pull/32

File: public/forms/validation/form.ttl
Comment (Windvis) [inline public/forms/validation/form.ttl:23]: Hmm, we need to tweak the property group setup in a future PR so that it doesn't show this title everywhere

File: public/forms/validation/form.ttl
Comment (Windvis) [inline public/forms/validation/form.ttl:155]:
```suggestion
            sh:path form:max;
```

It doesn't trigger the validation since the type is always set.

File: public/forms/validation/form.ttl
Comment (Windvis) [inline public/forms/validation/form.ttl:127]: I would give the concept a more descriptive uri instead of a UUID, or is this something that will be used outside of the form-builder as well?

File: .gitignore
Comment (JonasVanHoof) [reply .gitignore:5]: Didn't know that NICE

File: app/components/add-validations-to-form.js
Comment (JonasVanHoof) [reply app/components/add-validations-to-form.js:21]: It's just to make it clear that these are not variables that.. yeah I can change it so just camelCasing :sweat_smile:

Comment (Windvis) [reply app/components/add-validations-to-form.js:21]: Yea, in that case I would make them const variables in the module scope, or simply use the strings in the function directly 😄.

File: app/templates/formbuilder/edit.hbs
Comment (JonasVanHoof) [reply app/templates/formbuilder/edit.hbs:15]: Indeed putting in time and effort now to make it work is a bit of a waste. On Monday I also sat down with Marie and there we said that we are going to take a look on how we are going to organise everything within the form builder design wise & placements.

File: public/forms/validation/form.ttl
Comment (JonasVanHoof) [reply public/forms/validation/form.ttl:23]: https://github.com/lblod/frontend-form-builder/issues/33

File: public/forms/validation/form.ttl
Comment (JonasVanHoof) [reply public/forms/validation/form.ttl:155]: Indeed I now see the difference :D
image

File: public/forms/validation/form.ttl
Comment (JonasVanHoof) [reply public/forms/validation/form.ttl:127]: I could change it in the meta.ttl file as I only found [results on this ID](https://github.com/search?q=org%3Alblod%20dde3d2a3-e848-47ea-ba44-0f2e565f04ab&type=code) in the `form-builder` itself.

Does this also mean that we will be having  to create a migration for these as they are added to the database in the [backend](https://github.com/lblod/app-form-builder/blob/7e74170f0a26932a5fc58667919a865eb70dc642/config/migrations/20210305123324-add-validations.sparql#L11)?

File: app/components/add-validations-to-form.js
Comment (JonasVanHoof) [reply app/components/add-validations-to-form.js:71]: Directly created the builderForm in the callback method when creating the builderStore

File: app/components/add-validations-to-form.js
Comment (JonasVanHoof) [reply app/components/add-validations-to-form.js]: Created a `getter` for it + renamed it to `canShowRdfForm`

File: public/forms/validation/form.ttl
Comment (Windvis) [reply public/forms/validation/form.ttl:127]: Maybe, I'm not sure why we inserted these in the DB actually. Don't we only need them in the frontend?

Comment (JonasVanHoof) [reply public/forms/validation/form.ttl:127]: Than this was a good question :D It actually confused me a bit because when changing the objects in the `meta.ttl` (for example: the `form:required` in meta.ttl  now was a full uri before) file it actually just worked in the frontend so that made me think that it is indeed not needed that it is stored in the database. 

Another thing thing is when starting with the validation form @aatauil  got me quering the database and putting that ttl result in the valdiations meta ttl file. This again confirms that we do not need to write migrations for it?

Ill change the id to something more descriptive :)

File: app/styles/_shame.scss
Comment (JonasVanHoof) [reply app/styles/_shame.scss]: The buttons disappeared when adding more fields but added a margin now for the `au-o-box` in the playground that fixes it for the builder form and the validation form 
```css
.au-o-box {
  margin-bottom: 4rem;
}

Comment (Windvis) [reply app/styles/_shame.scss]: Hmmmmmmmmm 😄, I wouldn't extend Appuniversum classes like that since it will also affect other cases (or future cases we might add). We probably can fix it by restructuring the html setup, but as a quick fix I would use one of the margin spacing utils: https://appuniversum.github.io/ember-appuniversum/?path=/story/utilities-spacing--page

File: public/forms/validation/form.ttl
Comment (JonasVanHoof) [reply public/forms/validation/form.ttl]: Changed it to `http://lblod.data.gift/concept-schemes/possibleValidations`

File: app/styles/_shame.scss
Comment (JonasVanHoof) [reply app/styles/_shame.scss]: Woeps haha

It is fixed with the spacing style on the builder form but not yet fully happy about it in the validation form. But it "workable"

File: app/controllers/formbuilder/edit.js
Comment (JonasVanHoof) [inline app/controllers/formbuilder/edit.js:42]: Also try to use the derived state here as [mentioned](https://github.com/lblod/frontend-form-builder/pull/22#discussion_r1344071910)

**Review by Windvis — APPROVED** [review]

File: app/controllers/formbuilder/edit.js
Comment (Windvis) [reply app/controllers/formbuilder/edit.js:42]: This doesn't seem to be used (yet)? Does this block the merging of this PR?

Comment (JonasVanHoof) [reply app/controllers/formbuilder/edit.js]: Definitly not :D

**Comment by JonasVanHoof** [comment]
@Windvis im not merging it yet because I saw that not all the data is shown in the fields when switching between builder and validations tab. Also when changing somehting in the validation form there is not possibily to `Bewaar` the current changes

**Comment by Windvis** [comment]
Whoops, my bad. I thought you meant I could merge it 😬

**Comment by Windvis** [comment]
> Also when changing somehting in the validation form there is not possibily to Bewaar the current changes

I guess we also need to modify the `formChanged` value in the `onUpdateValidations` handler?

**Pull request merged**
